tests/shared/core/training_data/test_visualization.py

```diff
@@ -188,3 +190,42 @@ def test_story_visualization_with_merging(domain: Domain):
     assert 15 < len(generated_graph.nodes()) < 33
 
     assert 20 < len(generated_graph.edges()) < 33
+
+
+@pytest.mark.parametrize(
+    "input_nodes, input_edges, remove_count, expected_nodes, expected_edges",
+    [
+        (
+            [-2, -1, 0, 1, 2, 3, 4, 5],
+            [(-2, 0), (-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)],
+            3,
+            set([0, 1, 2, 3, 4, 5, -1]),
+            [(-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]
+        ),
+        (
+            [-3, -2, -1, 0, 1, 2, 3, 4, 5],
+            [(-3, -2), (-2, -1), (-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)],
+            4,
+            set([-3, -1, 0, 1, 2, 3, 4, 5]),
+            [(-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]
+        )
```

Comment:
You only have test cases when the graph is connected with only one edge between subsequent nodes. While the original function will remove all the predecessors nodes if it was already in `predecessors_seen` the refactored version will just remove the first one and continue. Please keep the same functionality after the refactor 🙏

Reply:
Hey @Urkem,
I have removed the break statement from the loop to ensure that all predecessor nodes present in the "predecessors_seen" collection are removed without exiting the loop. Now, This refactor maintains the same functionality as before. Please review it :)
Thank You!